File: Data-Structure/LeetCode_After_400_Done/Top Facebook Questions/42. Trapping Rain Water/trap.js
/*
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1] Output: 6

具体来说，仅仅对于位置i，能装下多少水呢？能装 2 格水，因为height[i]的高度为 0，而这里最多能盛 2 格水，2-0=2。 为什么位置i最多能盛 2 格水呢？因为，位置i能达到的水柱高度和其左边的最高柱子、右边的最高柱子有关，我们分别称这两个柱子高度为l_max和r_max；位置 i 最大的水柱高度就是min(l_max, r_max)。
# 左边最高的柱子,# 右边最高的柱子
water[i] = min(max(height[0..i]), max(height[i..end])) - height[i]
*/
{
    // 暴力算法：
    var trap = function (height) {
        let N = height.length;
        let res = 0;
        for (let i = 1; i < N - 1; i++) {
            let l_max = 0, r_max = 0;
            for (let j = i; j < N; j++)
                r_max = Math.max(r_max, height[j]);// 找右边最高的柱子
            for (let j = i; j >= 0; j--)
                l_max = Math.max(l_max, height[j]);// 找左边最高的柱子   
            res += Math.min(l_max, r_max) - height[i];
        }
        return res;
    }
}
{
    //备忘录优化
    var trap = function (height) {
        let N = height.length, res = 0, l_max = [], r_max = [];
        l_max[0] = height[0];
        r_max[N - 1] = height[N - 1];

        for (let i = 1; i < N; i++)
            l_max[i] = Math.max(height[i], l_max[i - 1]);// 从左向右计算 l_max
        for (let i = N - 2; i >= 0; i--)
            r_max[i] = Math.max(height[i], r_max[i + 1]); // 从右向左计算 r_max   
        for (let i = 1; i < N - 1; i++)
            res += Math.min(l_max[i], r_max[i]) - height[i];

        return res;
    }
}
{
    //双指针解法
    var trap = function (height) {
        let N = height.length;
        let res = 0;
        let l_max = height[0], r_max = height[N - 1];
        let left = 0, right = N - 1;

        while (left <= right) {
            l_max = Math.max(l_max, height[left]);
            r_max = Math.max(r_max, height[right]);

            // res += Math.min(l_max, r_max) - height[i]
            if (l_max < r_max) {
                res += l_max - height[left];
                left++;
            } else {
                res += r_max - height[right];
                right--;
            }
        }

        return res;
    }
}

File: Data-Structure/_Algorithm (LeetCode)/Alg-Study-Plan_Algorithm III/Day 4 Two Pointers/00042. Trapping Rain Water/trap.java
//42. Trapping Rain Water
/*
基本思路
PS：这道题在《算法小抄》 的第 364 页。

对于任意一个位置 i，能够装的水为：

Copy
water[i] = min(
           # 左边最高的柱子
           max(height[0..i]),
           # 右边最高的柱子
           max(height[i..end])
        ) - height[i]

关键在于，如何能够快速计算出某一个位置左侧所有柱子的最大高度和右侧所有柱子的最大高度。

这道题的解法比较多样，可以预计算数组，可以用 双指技巧，可以用单调栈技巧，这里就说一个最简单的解法，用预计算的方式求解，优化暴力解法的时间复杂度，更多解法请看详细题解。

详细题解：手把手搞懂接雨水问题的多种解法

标签：数组双指针

解法代码
*/
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) {
            return 0;
        }
        int n = height.length;
        int res = 0;
        // 数组充当备忘录
        int[] l_max = new int[n];
        int[] r_max = new int[n];

        // 初始化 base case
        l_max[0] = height[0];
        r_max[n - 1] = height[n - 1];

        // 从左向右计算 l_max
        for (int i = 1; i < n; i++)
            l_max[i] = Math.max(height[i], l_max[i - 1]);

        // 从右向左计算 r_max
        for (int i = n - 2; i >= 0; i--)
            r_max[i] = Math.max(height[i], r_max[i + 1]);

        // 计算答案
        for (int i = 1; i < n - 1; i++)
            res += Math.min(l_max[i], r_max[i]) - height[i];

        return res;
    }
}
